File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null){
            return false;
        }
        if (subRoot == null){
            return true;
        }

        //we basically go until the subRoot is null because that means all the previous ones we've checked pass the matching

        if (isSameTree(root, subRoot)){
            return true;//initial check
        }
        //actual method of recursing
        //if the left part of our current node or right part of our current node matches the subtree return true
        return (isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot));


    }
    public boolean isSameTree(TreeNode root, TreeNode subRoot){
        //if both trees are null, theyre both matching
        if (root == null && subRoot == null){
            return true;
        }
        //if both are not null but both values match then we go thru left and right nodes of both so that we can determine if they truly match or not
        if (root != null && subRoot != null && root.val == subRoot.val){
            return isSameTree(root.left, subRoot.left) && isSameTree(root.right, subRoot.right);
        }
        return false;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        #output in terms of [bool, val] [balanced, height]
        def dfs(root):
            if not root:
                return [True, 0] #base case is that its at the bottom node and its true
            left, right = dfs(root.left), dfs(root.right)
            balanced = left[0] and right[0] and abs(left[1]-right[1]) <= 1
            return [balanced, 1 + max(left[1], right[1])] #current node height
        return dfs(root)[0]

        
        